### ISSUE TITLE
Regarding future refactoring

### ISSUE THREAD

**Issue by daiplusplus** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#439]
@sjh37 in [a recent comment](https://github.com/sjh37/EntityFramework-Reverse-POCO-Code-First-Generator/pull/432#issuecomment-393129439) you remarked:

> In the EF.Core branch, I've done extensive refactoring into a project named 'Generator'. This is a plain C# DLL which will be used to develop the tt going forwards. Not yet, but I plan as part of the build process (via cake or something similar) that the .tt and .ttinclude files to be completely created from the 'Generator' project.

I have some thoughts regarding the refactoring: I can tell this project mutated from very simple beginnings, this can be seen with the inconsistent semantics of the `Settings` callbacks, for example, and how they're used (e.g. `TableRename`' is called for each table and the results are not cached, while `ForeignKeyProcessing` is passed in the whole set of foreign-keys for processing).

So I propose that any refactoring work should be towards a more "unified" design with a straightforward pipeline:

1. Script reads database metadata into an in-memory model, but does no additional processing, i.e. a "first draft" of the model.
2. Script then calls back to a single user-defined callback where the users' code performs table/column renames, column type changes, declaring views' keys, etc.
3. Script takes this "second draft" model and then generates an in-memory representation of the final output code (i.e. a first-draft of the view-model).
4. Script then calls back to a single user-defined callback where the view-model can be adjusted (this is where C# things like access-modifiers, namespaces, and such can be set.
5. Script then writes it out to files.

Something like:

    Database.tt: 
    <#@ include file="...ttinclude" #>
    <#
    var model = ReadDatabase( connectionString ); // included
    UserDefinedUpdateModel( model ); // defined below
    var viewModel = ToCSharp( model ); // included
    UserDefinedUpdateViewModel( viewModel ); // defined below
    Render( viewModel ); // included
    #>
    <#+
    static void UserDefinedUpdateModel( Model model ) {
        
         foreach( Table table in model.Tables ) {
              // rename table, set view primary key, etc
         }
    }

    static void UserDefinedUpdateViewModel( ViewModel vm ) {
        
         EntityClass customerClass = vm.EntityClasses.Single( ec => ec.Name =="Customer" );
         customerClass.AccessModifiers = "public sealed";
    }
    #>

**Comment by sjh37** [comment]
Yes, this project has grown over time, and now is the time I to completely re-write and refactor it to make a v3 which will support both ef6/7 and in the future multiple databases. It needs a serious looking at as I don't like how its grown over the years in a non unified way. I really like your ideas above, and you've obviously got a good developer head on your shoulders.
Do you have an email address? Ping me an email at simon [at] hicrest.net

**Comment by daiplusplus** [comment]
I did email you when you posted your comment reply - did you get it?

**Comment by sjh37** [comment]
The refactor has largely happened: v3 moved generation into the Generator project, and v4 moved schema reading into the `efrpg` tool. Replacing the separate callbacks with a single model callback would break every existing `Database.tt`. Closing.
